File: src/apps/daemon/internal/networking/transport/window_clamper.go
package transport

type WindowClamper struct {
	ClampedSize uint16
	Active      bool
}

func NewWindowClamper(clampedSize uint16) *WindowClamper {
	return &WindowClamper{
		ClampedSize: clampedSize,
		Active:      true,
	}
}

func (w *WindowClamper) ClampWindow(origWin uint16, isHandshake bool) uint16 {
	if !w.Active {
		return origWin
	}
	if isHandshake && origWin > w.ClampedSize {
		return w.ClampedSize
	}
	return origWin
}

func (w *WindowClamper) IsRstLegitimate(rstSeq uint32, lastAckSeq uint32, window uint32) bool {
	diff := rstSeq - lastAckSeq
	maxWin := window
	if maxWin < 4096 {
		maxWin = 4096
	}
	return diff <= maxWin
}

File: src/packages/control-ui/src/api/window_clamper.ts
export class WindowClamper {
  constructor(public clampedSize: number = 4, public enabled: boolean = true) {}

  clampWindow(originalWindow: number, isHandshake: boolean): number {
    if (!this.enabled) return originalWindow;
    return isHandshake && originalWindow > this.clampedSize ? this.clampedSize : originalWindow;
  }

  isRstLegitimate(rstSeq: number, lastAckSeq: number, window: number): boolean {
    const diff = rstSeq - lastAckSeq;
    const maxWin = Math.max(window, 4096);
    return diff >= 0 && diff <= maxWin;
  }
}
